guides/keras_cv/object_detection_keras_cv.py

```diff
@@ -0,0 +1,750 @@
+    optimizer=optimizer,
+    metrics=[coco_metrics],
+)
+result = model.evaluate(eval_ds)
+result = coco_metrics.result(force=True)
+print(tablify(result))
+
+"""
+**A note on TPU compatibility**
+
+Evaluation of `BoxCOCOMetrics` require running `tf.image.non_max_suppression()`
+inside of the `model.train_step()` and `model.evaluation_step()` functions.
+Due to this, the metric suite is not compatible with TPU when used with the
+`compile()` API.
+
+Luckily, there are two workarounds that allow you to still train a RetinaNet on TPU:
+
+- The use of a custom callback
+- Using a [SideCarEvaluator](https://www.tensorflow.org/api_docs/python/tf/keras/utils/SidecarEvaluator)
+
+Let's use a custom callback to achieve TPU compatibility in this guide:
+"""
+
+
+class EvaluateCOCOMetricsCallback(keras.callbacks.Callback):
+    def __init__(self, data):
+        super().__init__()
+        self.data = data
+        self.metrics = keras_cv.metrics.BoxCOCOMetrics(
+            bounding_box_format="xywh",
+            # passing 1e9 ensures we never evaluate until
+            # `metrics.result(force=True)` is
+            # called.
+            evaluate_freq=1e9,
+        )
+
+    def on_epoch_end(self, epoch, logs):
+        self.metrics.reset_state()
+        for batch in tqdm.tqdm(self.data):
+            images, y_true = batch[0], batch[1]
+            y_pred = self.model.predict(images, verbose=0)
+            self.metrics.update_state(y_true, y_pred)
+
+        metrics = self.metrics.result()
+        logs.update(metrics)
+        return logs
+
+
+"""
+## Training our model
+
+All that is left to do is train our model.  KerasCV object detection models follow the
+standard Keras workflow, leveraging `compile()` and `fit()`.
+
+Let's compile our model:
+"""
+
+model.compile(
+    classification_loss="focal",
+    box_loss="smoothl1",
+    optimizer=optimizer,
+    # We will use our custom callback to evaluate COCO metrics
+    metrics=None,
+)
+
+"""
+And run `model.fit()`!
+"""
+
+# If you want to train the fully model, uncomment `.take(20)` from each
+# of the following dataset references.
+
+model.fit(
+    train_ds.take(20),
+    validation_data=eval_ds.take(20),
+    # Run for 10-35~ epochs to achieve good scores.
+    epochs=1,
+    callbacks=[EvaluateCOCOMetricsCallback(eval_ds.take(20))],
+)
+
+"""
+To achieve reasonable scores, you will want to run `fit()` with 10-35~ epochs.
+"""
+
+"""
+Visualizing
+"""
+
+"""
+This metric can be used with `fit()` like any other callback, simply pass:
+`callbacks=callbacks + [EvaluateMetricsCallback(data=eval_ds)]` to your `fit()` call.
+
+## Inference and Plotting results
+
+KerasCV makes object detection inference simple.  `model.predict(images)` returns a
+RaggedTensor of bounding boxes.  By default, `RetinaNet.predict()` will perform
+a non max suppression operation for you.
+
+In this section, we will use a `keras_cv` provided preset:
+"""
+model = keras_cv.models.RetinaNet.from_preset(
```

Comment:
Didn't we already define this on line 58?

Reply:
Yes, but by doing this we re-initialize the model to make sure we have a fully pretrained model.  Above we override it, and then we don't get strong convergence necessarily.

Reply:
Should we call it `pretrained_model`?